[ValueTracking] Return true for AddrSpaceCast in canCreateUndefOrPoison

wenju-he opened the pull request:
In our downstream GPU target, following IR is valid before instcombine although the second addrspacecast causes UB.
  define i1 @test(ptr addrspace(1) noundef %v) {
    %0 = addrspacecast ptr addrspace(1) %v to ptr addrspace(4)
    %1 = call i32 @llvm.xxxx.isaddr.shared(ptr addrspace(4) %0)
    %2 = icmp eq i32 %1, 0
    %3 = addrspacecast ptr addrspace(4) %0 to ptr addrspace(3)
    %4 = select i1 %2, ptr addrspace(3) null, ptr addrspace(3) %3
    %5 = icmp eq ptr addrspace(3) %4, null
    ret i1 %5
  }
We have a custom optimization that replaces invalid addrspacecast with poison, and IR is still valid since `select` stops poison propagation.

However, instcombine pass optimizes `select` to `or`:
    %0 = addrspacecast ptr addrspace(1) %v to ptr addrspace(4)
    %1 = call i32 @llvm.xxxx.isaddr.shared(ptr addrspace(4) %0)
    %2 = icmp eq i32 %1, 0
    %3 = addrspacecast ptr addrspace(1) %v to ptr addrspace(3)
    %4 = icmp eq ptr addrspace(3) %3, null
    %5 = or i1 %2, %4
    ret i1 %5
The transform is invalid for our target.

wenju-he committed:
[InstCombine] Don't folder select to or if value argument is user of invalid addrspacecast inst

In our downstream GPU target, following IR is valid before instcombine
although the second addrspacecast causes UB.
  define i1 @test(ptr addrspace(1) noundef %v) {
    %0 = addrspacecast ptr addrspace(1) %v to ptr addrspace(4)
    %1 = call i32 @llvm.xxxx.isaddr.shared(ptr addrspace(4) %0)
    %2 = icmp eq i32 %1, 0
    %3 = addrspacecast ptr addrspace(4) %0 to ptr addrspace(3)
    %4 = select i1 %2, ptr addrspace(3) null, ptr addrspace(3) %3
    %5 = icmp eq ptr addrspace(3) %4, null
    ret i1 %5
  }
We have a custom optimization that replaces invalid addrspacecast with
poison, and IR is still valid since `select` stops poison propagation.

However, instcombine pass optimizes `select` to `or`:
    %0 = addrspacecast ptr addrspace(1) %v to ptr addrspace(4)
    %1 = call i32 @llvm.xxxx.isaddr.shared(ptr addrspace(4) %0)
    %2 = icmp eq i32 %1, 0
    %3 = addrspacecast ptr addrspace(1) %v to ptr addrspace(3)
    %4 = icmp eq ptr addrspace(3) %3, null
    %5 = or i1 %2, %4
    ret i1 %5
The transform is invalid for our target.

nikic reviewed (changes requested):
If an invalid addrspacecast can produce poison, then this must be specified in https://github.com/llvm/llvm-project/blob/34a48941498d95ec2682f7adaeb6115b7b4d70ba/llvm/lib/Analysis/ValueTracking.cpp#L7352, not here. LangRef should also be adjusted in https://llvm.org/docs/LangRef.html#addrspacecast-to-instruction to specify that addrspacecast can introduce poison.

wenju-he committed:
update LangRef, refine test

wenju-he commented:
> If an invalid addrspacecast can produce poison, then this must be specified in
> 
> https://github.com/llvm/llvm-project/blob/34a48941498d95ec2682f7adaeb6115b7b4d70ba/llvm/lib/Analysis/ValueTracking.cpp#L7352

This PR uses isValidAddrSpaceCast query from TTI which is available in InstCombine. But TTI isn't available in above canCreateUndefOrPoison function.
If we unconditionally return true for addrspacecast in canCreateUndefOrPoison, there is regression in following test probably because canCreateUndefOrPoison returning true for valid addrspacecast is preventing the optimization: https://github.com/llvm/llvm-project/blob/5875fafdc547889fb089c943a881a9ab6d8a23c0/llvm/test/Transforms/Attributor/reduced/aapointer_info_map_invalidation.ll#L4-L19

> LangRef should also be adjusted in https://llvm.org/docs/LangRef.html#addrspacecast-to-instruction to specify that addrspacecast can introduce poison.

done in ffff2c3624239c1290e71abcce44f3317aab597a, please review

wenju-he commented on llvm/test/Transforms/InstCombine/AMDGPU/addrspacecast.ll:
```diff
@@ -0,0 +1,23 @@
+; CHECK-NEXT:    ret i1 [[TMP4]]
+;
+  entry:
+  %0 = addrspacecast ptr addrspace(1) %x to ptr
```
done

wenju-he commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3194,8 +3194,23 @@ static Instruction *foldNestedSelects(SelectInst &OuterSelVal,
-                                bool Expected) {
+static bool impliesPoisonOrCond(
+    const Value *ValAssumedPoison, const Value *V, bool Expected,
+    llvm::function_ref<bool(unsigned, unsigned)> isValidAddrSpaceCast) {
```
done

dtcxzyw commented:
> > If an invalid addrspacecast can produce poison, then this must be specified in
> > https://github.com/llvm/llvm-project/blob/34a48941498d95ec2682f7adaeb6115b7b4d70ba/llvm/lib/Analysis/ValueTracking.cpp#L7352
> 
> This PR uses isValidAddrSpaceCast query from TTI which is available in InstCombine. But TTI isn't available in above canCreateUndefOrPoison function. If we unconditionally return true for addrspacecast in canCreateUndefOrPoison, there is regression in following test probably because canCreateUndefOrPoison returning true for valid addrspacecast is preventing the optimization:
> 
> https://github.com/llvm/llvm-project/blob/5875fafdc547889fb089c943a881a9ab6d8a23c0/llvm/test/Transforms/Attributor/reduced/aapointer_info_map_invalidation.ll#L4-L19
> 
> > LangRef should also be adjusted in https://llvm.org/docs/LangRef.html#addrspacecast-to-instruction to specify that addrspacecast can introduce poison.
> 
> done in [ffff2c3](https://github.com/llvm/llvm-project/commit/ffff2c3624239c1290e71abcce44f3317aab597a), please review

I don't think special handling of addrspacecast in `logical and/or->bitwise and/or` works. There are many places which call `isGuaranteedNotToBeUndefOrPoison`. For example, eliminating freeze from `freeze ptr (addrspacecast ptr addrspace(5) noundef %x to ptr)` is still invalid. That is, I can construct an input to bypass the workaround and cause the miscompilation.
It would be better to default to returning true for addrspacecast in `canCreateUndefOrPoison`, then add an optional callback for `isValidAddrSpaceCast` queries.

nikic commented on llvm/docs/LangRef.rst:
```diff
@@ -12621,6 +12621,9 @@ have no side effects, and must not capture the value of the pointer.
 :ref:`poison <poisonvalues>`.
 
+If the source is not :ref:`poison <poisonvalues>`, and the result pointer is
+non-dereferenceable, the result is :ref:`poison <poisonvalues>`.
```
This is incorrect, not every addrspacecast of a non-dereferenceable pointer should automatically result in poison (e.g. the pointer to the end of an object is non-dereferenceable, but should certainly not turn into poison).

This should instead say something along the lines of:

> Which address space casts are supported depends on the target. Unsupported address space casts return a poison value.

wenju-he replied:
done in 3128b338492a977b35b7c39bc770e85a703ca8d0, thanks @nikic

wenju-he committed:
return true for AddrSpaceCast in canCreateUndefOrPoison, update AttributorAttributes to restore previous behavior in aapointer_info_map_invalidation.ll

wenju-he commented:
> > > If an invalid addrspacecast can produce poison, then this must be specified in
> > > https://github.com/llvm/llvm-project/blob/34a48941498d95ec2682f7adaeb6115b7b4d70ba/llvm/lib/Analysis/ValueTracking.cpp#L7352
> > 
> > 
> > This PR uses isValidAddrSpaceCast query from TTI which is available in InstCombine. But TTI isn't available in above canCreateUndefOrPoison function. If we unconditionally return true for addrspacecast in canCreateUndefOrPoison, there is regression in following test probably because canCreateUndefOrPoison returning true for valid addrspacecast is preventing the optimization:
> > https://github.com/llvm/llvm-project/blob/5875fafdc547889fb089c943a881a9ab6d8a23c0/llvm/test/Transforms/Attributor/reduced/aapointer_info_map_invalidation.ll#L4-L19
> > 
> > > LangRef should also be adjusted in https://llvm.org/docs/LangRef.html#addrspacecast-to-instruction to specify that addrspacecast can introduce poison.
> > 
> > 
> > done in [ffff2c3](https://github.com/llvm/llvm-project/commit/ffff2c3624239c1290e71abcce44f3317aab597a), please review
> 
> I don't think special handling of addrspacecast in `logical and/or->bitwise and/or` works. There are many places which call `isGuaranteedNotToBeUndefOrPoison`. For example, eliminating freeze from `freeze ptr (addrspacecast ptr addrspace(5) noundef %x to ptr)` is still invalid. That is, I can construct an input to bypass the workaround and cause the miscompilation. It would be better to default to returning true for addrspacecast in `canCreateUndefOrPoison`, then add an optional callback for `isValidAddrSpaceCast` queries.

thanks @dtcxzyw I've changed to return true for addrspacecast in `canCreateUndefOrPoison` in 3128b338492a977b35b7c39bc770e85a703ca8d0. Adding an optional callback looks like the same as querying isValidAddrSpaceCast right after canCreateUndefOrPoison call, so I added isValidAddrSpaceCast query in llvm/lib/Transforms/IPO/AttributorAttributes.cpp to restore previous behavior of llvm/test/Transforms/Attributor/reduced/aapointer_info_map_invalidation.ll. The test is also updated to specify triple so that TTI is available.

Does it make sense to add an optional TTI argument to canCreateUndefOrPoison/isGuaranteedNotToBePoison/isGuaranteedNotToBeUndefOrPoison? If yes, then we don't need to query isValidAddrSpaceCast right after isGuaranteedNotToBeUndefOrPoison call in AttributorAttributes and potentially some other places.

wenju-he committed:
update clang tests

nikic commented:
For the purposes of this PR, I think you should only change canCreateUndefOrPoison to return true and just eat the regressions. We can follow up with using isValidAddrSpaceCast() in a followup, because it will be less straightforward. I'm not willing to accept direct use of TTI in ValueTracking without laundering it through an abstraction first (see https://discourse.llvm.org/t/constant-propagation-for-target-specific-intrinsics/85881/5).

wenju-he committed:
revert change to AttributorAttributes

nikic commented on llvm/test/Transforms/InstCombine/AMDGPU/addrspacecast.ll:
```diff
@@ -0,0 +1,24 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=instcombine %s | FileCheck %s
+
+; Check that `select B, true, C` isn't optimized to `or B, C`.
```
```suggestion
; Check that `select B, true, C` isn't optimized to `or B, C`,
; because the invalid addrspacecast %asc.shared may introduce poison.
```

nikic reviewed (approved):
LGTM

wenju-he committed:
Update llvm/test/Transforms/InstCombine/AMDGPU/addrspacecast.ll

Co-authored-by: Nikita Popov <github@npopov.com>

arsenm commented on llvm/test/Transforms/InstCombine/AMDGPU/addrspacecast.ll:
```diff
@@ -0,0 +1,25 @@
+; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=instcombine %s | FileCheck %s
+
+; Check that `select B, true, C` isn't optimized to `or B, C`,
+; because the invalid addrspacecast %asc.shared may introduce poison.
```
Description is slightly inaccurate. In this case it will always be poison

wenju-he replied:
> Description is slightly inaccurate. In this case it will always be poison

done

wenju-he committed:
remove 'may' in test comment
